ENH: scil_resample_tractogram

AntoineTheb opened the pull request:
Script to resample a tractogram to a specific number of streamlines (in the dumbest way possible). If the tractogram has more streamlines than needed, the tractogram is subsampled to the number specified. If the tractogram has less streamlines, the tractogram is "upsampled" with new streamlines by adding noise to existing streamlines. The "upsampled" tractogram can be smoothed afterwards. 

For example, here is a u-shaped, hard-to-track bundle on the DiSCo3 dataset, upsampled to 1000 streamlines:

![sub-DiSCo3_23](https://user-images.githubusercontent.com/10272382/127553135-14afb638-5553-4092-8efb-ede9f7842663.jpg)

Allows to compensate for over or under-represented bundles. Kinda did that on a whim in 5 minutes so let me know if this script is useless or wrong or already exists. 

```
usage: scil_resample_tractogram.py [-h]
                                   (--point_wise_std POINT_WISE_STD | --streamline_wise_std STREAMLINE_WISE_STD)
                                   [--gaussian SIGMA | --spline SIGMA NB_CTRL_POINT]
                                   [-e ERROR_RATE]
                                   [--keep_invalid_streamlines]
                                   [--reference REFERENCE] [-f] [-v]
                                   in_tractogram nb_streamlines out_tractogram

Script to resample a tractogram to a set number of streamlines.
Can be useful to build training sets for machine learning algorithms, to
upsample under-represented bundles or downsample over-represented bundles.

Works by either selecting a subset of streamlines or by generating new
streamlines by adding gaussian noise to existing ones.

Includes smoothing to compensate for the noisiness of new streamlines
generated by the process.

Example usage:
$ scil_resample_tractogram.py input.trk 1000 output.trk --point_wise_std 0.5 --spline 5 10 --keep_invalid_streamlines
$ scil_visualize_bundles.py output.trk --local_coloring --width=0.1

positional arguments:
  in_tractogram         Input tractography file.
  nb_streamlines        Number of streamlines to resample the tractogram to.
  out_tractogram        Output tractography file.

optional arguments:
  -h, --help            show this help message and exit
  --point_wise_std POINT_WISE_STD
                        Noise to add to existing streamlines'points to generate new ones.
  --streamline_wise_std STREAMLINE_WISE_STD
                        Noise to add to existing whole streamlines to generate new ones.
  --gaussian SIGMA      Sigma for smoothing. Use the value of surronding
                        X,Y,Z points on the streamline to blur the streamlines.
                        A good sigma choice would be around 5.
  --spline SIGMA NB_CTRL_POINT
                        Sigma for smoothing. Model each streamline as a spline.
                        A good sigma choice would be around 5 and control point around 10.
  -e ERROR_RATE         Maximum compression distance in mm after smoothing. [0.1]
  --keep_invalid_streamlines
                        Keep invalid newly generated streamlines that may go out of the bounding box.
  --reference REFERENCE
                        Reference anatomy for tck/vtk/fib/dpy file
                        support (.nii or .nii.gz).
  -f                    Force overwriting of the output files.
  -v                    If set, produces verbose output.
```

To test:

```bash
$ ./scripts/scil_resample_tractogram.py ~/braindata/databases/DiSCo/derivatives/bundles/output/sub-DiSCo3/sub-DiSCo3_23.trk 1000 sub-DiSCo3_23_resmapled.trk --point=0.5 -f --keep --gaussian 5 --seed 1337
$ scil_visualize_bundles.py sub-DiSCo3_23_resmapled.trk --local_coloring --width=0.1
```

The upsampling part could definitely be improved in the future to generate less noisy streamlines.

@arnaudbore @GuillaumeTh @mdesco

scil-admin commented:
Build passed ! Good Job :beers: !

AntoineTheb commented:
I realized a few days ago that it would be less messy to add the same "noise" to every point in the streamline instead of different noise to every point. This would essentially translate the streamline and thus give less noisy new streamlines. 

Will do once I have a few spare minutes.

frheault commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,75 @@
+    original_number = len(sft.streamlines)
+
+    if args.nb_streamlines > original_number:
+        sft = upsample_tractogram(sft, args.nb_streamlines, args.std)
```
You could smooth directly in there. Adding a smoothing parameter and smoothing streamlines directly after the point-wise noise.
Then you could apply the streamlines-wise noise.

frheault commented on scilpy/utils/streamlines.py:
```diff
@@ -664,3 +664,28 @@ def cut_invalid_streamlines(sft):
+    """
+    new_indices = np.random.choice(
+        range(len(sft.streamlines)), nb, replace=False)
+    return filter_tractogram_data(sft, new_indices)
```
sft[new_indices] should work

mdesco commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+""" Dumb script to resample a tractogram to a set number of streamlines.
```
Remove "Dumb" nothing is dumb ;)

mdesco commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+
+""" Dumb script to resample a tractogram to a set number of streamlines.
+Might be useful to build training sets for machine learning algorithms, to
```
Can be useful

mdesco commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,75 @@
+streamlines by adding gaussian noise to existing ones.
+
+It is recommended to use "scil_smooth_streamlines" afterwards to clean the
+noisy new streamlines.
```
Please provide your example 3 line execution as example, i.e. resample, smooth, visualize as in the PR conversation. This is really helpful for the future user

mdesco reviewed (approved):
Vraiment cool! I love it.

AntoineTheb commented on scilpy/utils/streamlines.py:
```diff
@@ -664,3 +664,28 @@ def cut_invalid_streamlines(sft):
+    indices = np.random.choice(
+        len(sft.streamlines), nb_new)
+    new_streamlines = sft.streamlines.copy()
+    for s in sft.streamlines[indices]:
```
Not sure how I could generate new streamlines using sft.streamlines._data tho. Will keep the loop.

pep8speaks commented:
Hello @AntoineTheb, Thank you for updating !



* In the file [`scripts/scil_resample_tractogram.py`](https://github.com/scilus/scilpy/blob/24d3e838be0e64d593e85292acd3ac69203c3590/scripts/scil_resample_tractogram.py):

> [Line 57:80](https://github.com/scilus/scilpy/blob/24d3e838be0e64d593e85292acd3ac69203c3590/scripts/scil_resample_tractogram.py#L57): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

AntoineTheb commented:
@frheault @mdesco @arnaudbore Addressed comments in 24d3e838be0e64d593e85292acd3ac69203c3590 and a0ef426742e278536ca1b037d15d3e05990523e2. Ready for re-review.

scil-admin commented:
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/2/

scil-admin commented:
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/3/

scil-admin commented:
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-485/4/

scil-admin commented:
Build passed ! Good Job :beers: !

GuillaumeTh commented:
I think this script to do much things. We already have a script that downsample streamlines: scil_get_subset_streamlines.py. However, I really like the script to generate streamlines by adding noise.

AntoineTheb commented on scilpy/utils/streamlines.py:
```diff
@@ -3,14 +3,15 @@
 import itertools
 from functools import reduce
 import logging
+import numpy as np
```
Fixed

scil-admin commented:
Build passed ! Good Job :beers: !

arnaudbore commented on scilpy/utils/streamlines.py:
```diff
@@ -664,3 +665,40 @@ def cut_invalid_streamlines(sft):
+    streamline_wise_std=None, gaussian=None, spline=None, seed=None
+):
+    """ Generate new streamlines to add to a tractogram samping a gaussian
+    centered around multiple streamlines existing points.
```
clean docstring

AntoineTheb replied:
Cleaned up and reworded in df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

arnaudbore commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,125 @@
+                            'spline.\nA good sigma choice would be around 5 '
+                            'and control point around 10.')
+
+    p.add_argument('-e', dest='error_rate', type=float, default=0.1,
```
-e or --e  ?

arnaudbore replied:
Please remove the compressing.

AntoineTheb replied:
Donesies in df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

arnaudbore commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,125 @@
+    elif args.nb_streamlines < original_number:
+        sft = get_subset_streamlines(sft, args.nb_streamlines, args.seed)
+    streamlines = compress_streamlines(
+        sft.streamlines, args.error_rate)
```
I would not use compressing within this script.

AntoineTheb replied:
Yeah me neither df4d41a44f3f2bfd5fcf3c4899142cfeccbc5675

arnaudbore reviewed (approved):
Last comments otherwise LGTM

scil-admin commented:
Build passed ! Good Job :beers: !

scil-admin commented:
Build passed ! Good Job :beers: !

arnaudbore commented on scilpy/utils/streamlines.py:
```diff
@@ -664,3 +665,41 @@ def cut_invalid_streamlines(sft):
+    """ Generate new streamlines by either adding gaussian noise around
+    streamlines' points, or by translating copies of existing streamlines
+    by a random amount.
+    """
```
clean docstring

AntoineTheb replied:
Clean quoi ?

GuillaumeTh replied:
I think it means to add a description for each parameters.

AntoineTheb replied:
Donesoes in 93351ff

arnaudbore commented on scripts/tests/test_resample_tractogram.py:
```diff
@@ -12,14 +12,15 @@
+                             'union_shuffle_sub.trk')
+    ret = script_runner.run('scil_resample_tractogram.py', in_tracto,
+                            '1000', 'union_shuffle_sub_resampled.trk',
+                            '--point_wise_std', '0.5')
```
add a test using upsampling

AntoineTheb replied:
Tests were reworked in 93351ff

arnaudbore commented on scripts/scil_resample_tractogram.py:
```diff
@@ -0,0 +1,117 @@
+    std_group = p.add_mutually_exclusive_group()
+    std_group.add_argument('--point_wise_std', type=float,
+                           help='Noise to add to existing streamlines\'' +
+                                'points to generate new ones.')
```
```suggestion
                                ' points to generate new ones.')
```

scil-admin commented:
Build passed ! Good Job :beers: !
